File: Examples/ImageRegistrationMethodBSpline3/ImageRegistrationMethodBSpline3.py
#!/usr/bin/env python
# =========================================================================
#
#  Copyright NumFOCUS
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0.txt
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
# =========================================================================

import SimpleITK as sitk
import sys
import os


def command_iteration(method, bspline_transform):
    if method.GetOptimizerIteration() == 0:
        # The BSpline is resized before the first optimizer
        # iteration is completed per level. Print the transform object
        # to show the adapted BSpline transform.
        print(bspline_transform)

    print(f"{method.GetOptimizerIteration():3} = {method.GetMetricValue():10.5f}")


def command_multi_iteration(method):
    # The sitkMultiResolutionIterationEvent occurs before the
    # resolution of the transform. This event is used here to print
    # the status of the optimizer from the previous registration level.
    if R.GetCurrentLevel() > 0:
        print(f"Optimizer stop condition: {R.GetOptimizerStopConditionDescription()}")
        print(f" Iteration: {R.GetOptimizerIteration()}")
        print(f" Metric value: {R.GetMetricValue()}")

    print("--------- Resolution Changing ---------")


if len(sys.argv) < 4:
    print("Usage:", sys.argv[0], "<fixedImageFilter> <movingImageFile>",
          "<outputTransformFile>")
    sys.exit(1)

fixed = sitk.ReadImage(sys.argv[1], sitk.sitkFloat32)

moving = sitk.ReadImage(sys.argv[2], sitk.sitkFloat32)

transformDomainMeshSize = [2] * fixed.GetDimension()
tx = sitk.BSplineTransformInitializer(fixed,
                                      transformDomainMeshSize)

print(f"Initial Number of Parameters: {tx.GetNumberOfParameters()}")

R = sitk.ImageRegistrationMethod()
R.SetMetricAsJointHistogramMutualInformation()

R.SetOptimizerAsGradientDescentLineSearch(5.0,
                                          100,
                                          convergenceMinimumValue=1e-4,
                                          convergenceWindowSize=5)

R.SetInterpolator(sitk.sitkLinear)

R.SetInitialTransformAsBSpline(tx,
                               inPlace=True,
                               scaleFactors=[1, 2, 5])
R.SetShrinkFactorsPerLevel([4, 2, 1])
R.SetSmoothingSigmasPerLevel([4, 2, 1])

R.AddCommand(sitk.sitkIterationEvent, lambda: command_iteration(R, tx))
R.AddCommand(sitk.sitkMultiResolutionIterationEvent,
             lambda: command_multi_iteration(R))

outTx = R.Execute(fixed, moving)

print("-------")
print(tx)
print(outTx)
print(f"Optimizer stop condition: {R.GetOptimizerStopConditionDescription()}")
print(f" Iteration: {R.GetOptimizerIteration()}")
print(f" Metric value: {R.GetMetricValue()}")

sitk.WriteTransform(outTx, sys.argv[3])

if ("SITK_NOSHOW" not in os.environ):
    resampler = sitk.ResampleImageFilter()
    resampler.SetReferenceImage(fixed)
    resampler.SetInterpolator(sitk.sitkLinear)
    resampler.SetDefaultPixelValue(100)
    resampler.SetTransform(outTx)

    out = resampler.Execute(moving)
    simg1 = sitk.Cast(sitk.RescaleIntensity(fixed), sitk.sitkUInt8)
    simg2 = sitk.Cast(sitk.RescaleIntensity(out), sitk.sitkUInt8)
    cimg = sitk.Compose(simg1, simg2, simg1 // 2. + simg2 // 2.)
    sitk.Show(cimg, "Image Registration Composition")

File: Examples/ImageRegistrationMethodBSpline3/ImageRegistrationMethodBSpline3.R
#=========================================================================
#
#  Copyright NumFOCUS
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0.txt
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
#=========================================================================
# Run with:
#
# Rscript --vanilla ImageRegistrationMethodBSpline3.R fixedImageFilter movingImageFile outputTransformFile

library(SimpleITK)

# Callback invoked each iteration
command_iteration <- function(method, bspline_transform) {
  if (method$GetOptimizerIteration() == 0) {
    # The BSpline is resized before the first optimizer
    # iteration is completed per level. Print the transform object
    # to show the adapted BSpline transform.
    cat(bspline_transform$ToString(), "\n")
  }

  cat(sprintf("%3d = %10.5f\n",
              method$GetOptimizerIteration(),
              method$GetMetricValue()))
}

# Callback invoked before starting a multi-resolution level.
command_multi_iteration <- function(method) {
  if (method$GetCurrentLevel() > 0) {
    cat(sprintf("Optimizer stop condition: %s\n",
                method$GetOptimizerStopConditionDescription()))
    cat(sprintf(" Iteration: %d\n", method$GetOptimizerIteration()))
    cat(sprintf(" Metric value: %f\n", method$GetMetricValue()))
  }

  cat("--------- Resolution Changing ---------\n")
}

args <- commandArgs( TRUE )

if (length(args) < 3) {
    stop("Usage: fixedImageFilter movingImageFile outputTransformFile")
}

fixed <- ReadImage(args[1], 'sitkFloat32')
moving <- ReadImage(args[2], 'sitkFloat32')

transformDomainMeshSize <- rep(2, fixed$GetDimension())
tx <- BSplineTransformInitializer(fixed, transformDomainMeshSize)

cat(sprintf("Initial Number of Parameters: %d\n", tx$GetNumberOfParameters()))

R <- ImageRegistrationMethod()
R$SetMetricAsJointHistogramMutualInformation()

learningRate <- 5.0
numberOfIterations <- 100
convergenceMinimumValue <- 1e-4
convergenceWindowSize <- 5
R$SetOptimizerAsGradientDescentLineSearch(learningRate, numberOfIterations,
                                          convergenceMinimumValue, convergenceWindowSize)

R$SetInterpolator('sitkLinear')

R$SetInitialTransformAsBSpline(tx, inPlace=TRUE, scaleFactors=c(1, 2, 5))
R$SetShrinkFactorsPerLevel(c(4, 2, 1))
R$SetSmoothingSigmasPerLevel(c(4, 2, 1))

R$AddCommand('sitkIterationEvent', function() command_iteration(R, tx))
R$AddCommand('sitkMultiResolutionIterationEvent', function() command_multi_iteration(R))

outTx <- R$Execute(fixed, moving)

cat("-------\n")
cat(tx$ToString(), "\n")
cat(outTx$ToString(), "\n")
cat(sprintf("Optimizer stop condition: %s\n", R$GetOptimizerStopConditionDescription()))
cat(sprintf(" Iteration: %d\n", R$GetOptimizerIteration()))
cat(sprintf(" Metric value: %f\n", R$GetMetricValue()))

WriteTransform(outTx, args[3])

if (Sys.getenv("SITK_NOSHOW") == "") {
  resampler <- ResampleImageFilter()
  resampler$SetReferenceImage(fixed)
  resampler$SetInterpolator('sitkLinear')
  resampler$SetDefaultPixelValue(100)
  resampler$SetTransform(outTx)

  out <- resampler$Execute(moving)
  simg1 <- Cast(RescaleIntensity(fixed), 'sitkUInt8')
  simg2 <- Cast(RescaleIntensity(out), 'sitkUInt8')
  cimg <- Compose(simg1, simg2, simg1 %/% 2.0 + simg2 %/% 2.0)
  Show(cimg, "Image Registration Composition")
}
